restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/asyncclient/HttpAsycClientHelper.java

```diff
@@ -0,0 +1,236 @@
+        } else {
+            if (logger.isInfoEnabled()) {
+                logger.info("DownloaderResponse wont be send because sender is :" + (sender.isTerminated() ? "terminated" : "null"));
+            }
+        }
+    }
+    public HttpUriRequest request (final HttpRequestDescriptor descriptor) throws IllegalArgumentException, URISyntaxException,
+            UnsupportedEncodingException {
+        final URI uri = descriptor.getUri();
+        final String method = descriptor.getMethod();
+        HttpUriRequest httpUriRequest = null;
+        if ("GET".equalsIgnoreCase(method)) {
+            final String query = descriptor.getParametersAsString();
+            URI result = null;
+            if (query != null && !query.isEmpty()) {
+                result = new URIBuilder()
+                        .setScheme(uri.getScheme())
+                        .setHost(uri.getHost())
+                        .setPort(uri.getPort())
+                        .setPath(uri.getPath())
+                        .setQuery(query)
+                        .build();
+            } else {
+                result = uri;
+            }
+            httpUriRequest = new HttpGet(result);
+        } else if ("POST".equalsIgnoreCase(method)) {
+            final List<NameValuePair> parameters = descriptor.getParameters();
+            final HttpPost post = new HttpPost(uri);
+            post.setEntity(new UrlEncodedFormEntity(parameters, "UTF-8"));
+            httpUriRequest = post;
+        } else {
+            throw new IllegalArgumentException(method + " is not a supported downloader method.");
+        }
+        if(descriptor.getHeaders() != null && descriptor.getHeaders().length>0){
+            httpUriRequest.setHeaders(descriptor.getHeaders());
+        }
+        return httpUriRequest;
+    }
+
+    private HttpResponseDescriptor response (final HttpRequest request, final HttpResponse response) throws IOException {
+        final HttpResponseDescriptor.Builder builder = HttpResponseDescriptor.builder();
+        final URI uri = URI.create(request.getRequestLine().getUri());
+        builder.setURI(uri);
+        builder.setStatusCode(response.getStatusLine().getStatusCode());
+        builder.setStatusDescription(response.getStatusLine().getReasonPhrase());
+        builder.setHeaders(response.getAllHeaders());
+        final HttpEntity entity = response.getEntity();
+        if (entity != null) {
+            InputStream stream = entity.getContent();
+            try {
+                final Header contentEncoding = entity.getContentEncoding();
+                if (contentEncoding != null) {
+                    builder.setContentEncoding(contentEncoding.getValue());
+                }
+                final Header contentType = entity.getContentType();
+                if (contentType != null) {
+                    builder.setContentType(contentType.getValue());
+                }
+                builder.setContent(StringUtils.toString(stream));
```

Comment:
i dont like how response is stringified here, i think it should be original "sender" responsability to do this if appropiate. I can see the existing HttpResponseDescriptor  already expectsa  String content, but i think its a bit dangerous in terms of encoding(this is assuming body is JVM charset encoded/UTF8...) and performance(we loss any streaming chance, and we create an String in memory with potentially big body with no control on actual length)

not sure worth the refactoring, but very worrying...

Reply:
should we refactor this as part of this PR?

Reply:
it will impact Downloader as well,so I guess we can do that later...